staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go

```diff
@@ -318,3 +342,62 @@ func getEnvelopePrefixTransformer(config *apiserverconfig.KMSConfiguration, enve
 		Prefix:      []byte(prefix + config.Name + ":"),
 	}, nil
 }
+
+func isKMSPluginReady(config *apiserverconfig.KMSConfiguration) (bool, error) {
```

Comment:
Could this only return the error? If arg parsing fails _or_ if healthz polling fails, return a non-nil error

Reply:
Maybe, but I want to deffirentate between the scenarios 
1. Healthz could not be checked due to configuration errors - false, error
2. Heathz returned non 200 status - false,nil
3. Healthz return 200 - true,nil

I think this approach fits with the signature of the poll.Wait.

Reply:
Hmm, but shouldn't you retry even if healthz endpoint is unreachable?
For example when KMS plugin starts later than kube-apiserver.
If yes, then there's no difference between non-200 status and any request errors

Reply:
Let me step back.
wait.PollImmediate takes a condition function of the following signature:
type ConditionFunc func() (done bool, err error)
Therefore, getHelathz must match that signature.

I think the reason for that signature is that polling status has three potential states:
1. true - we are done ,no need to retry further (kms-plugin OK)
2. false - poll failed, keep retrying (kms-plugin is not up yet)
3. error - got negative status, no need to poll anymore (ex. missing permissions on the key)

Reply:
`getHealthz` doesn't have to conform to what `wait.PollImmediate` expects, you can always adapt it with a closure.

I understand the purpose of those separate return values, but my point is: getHealthz should never cause wait.PollImmediate to exit early with an error.
Currently you return non-nil error when response code is not 200. I think kube-apiserver should keep retrying in that case, let the plugin flip response code to 200 later on.
Then KMS plugin can say "i'm up, but not ready to serve requests yet" and kube-apiserver won't crashloop because of it.

Reply:
ACK on the signature, you are correct.

Here a scenario:
kube-apiserver polls kms-plugin and the plugin responds with 500 because KMS returned "Key Destroyed Error" (or other non-retriable error). Assuming we use the logic that your propose, kube-apiserver will continue to retry for the duration of the Poll timeout which seems unnecessary to me.

So I see a couple of choices here:
1. Retry until we get either true or poll timeout expires
2. Differentiate between 500(return an error) and 503 (return false, nil)

WDYT?

Reply:
Special-casing 500 and causing it to abort retries immediately sounds fine.
I'm not sure what that achieves though.
It'll cause kube-apiserver to crash and be restarted a few seconds later anyway.

Retries will continue indefinitely (indirectly through kube-apiserver restarts) in any case.

Reply:
@awly PTAL - added a case for 500.

I agree that this may not make a lot of difference. However, I don't want to make any assumptions about how kube-apiserver will be restarted (i.e. its restart policy) or whether or not kube-apiserver should crash if kms-plugin is down.
My objective is to stop polling if KMS-Plugin returned a non-retryable error.

Reply:
SGTM, as long as there's some way for KMS plugin implementors to discover this expectation from /healthz